File: ts/project/.sdk/tm/csharp/feature/RatelimitFeature.cs
// Client-side rate limiting via a token bucket. Each request consumes a
// token; when the bucket is empty the request waits until the bucket
// refills at `rate` tokens per second (with capacity `burst`, default:
// rate). This keeps the client under a server's published quota rather
// than discovering it via 429s. The clock (`now`) and the wait (`sleep`)
// are injectable so the accounting can be tested deterministically.

using static ProjectNameSdk.Feature.FeatureOptions;

namespace ProjectNameSdk.Feature;

public class RatelimitFeature : BaseFeature
{
    private ProjectNameSDK? _client;
    private Dictionary<string, object?>? _options;
    private double _tokens;
    private long _last;

    // Activity tracking (mirrors the ts client._ratelimit record).
    public int Throttled;
    public int WaitMs;

    public RatelimitFeature()
    {
        Version = "0.0.1";
        Name = "ratelimit";
        Active = true;
    }

    public override void Init(Context ctx, Dictionary<string, object?> options)
    {
        _client = ctx.Client;
        _options = options;
        Active = FoptBool(options, "active", false);

        if (!Active)
        {
            return;
        }

        var rate = FoptNum(_options, "rate", 5);
        var burst = FoptNum(_options, "burst", rate);
        _tokens = burst;
        _last = FoptNow(_options)();

        var inner = ctx.Utility!.Fetcher;

        ctx.Utility.Fetcher = (ctx2, url, fetchdef) =>
        {
            Acquire();
            return inner(ctx2, url, fetchdef);
        };
    }

    private void Acquire()
    {
        var rate = FoptNum(_options, "rate", 5);
        var burst = FoptNum(_options, "burst", rate);

        // Refill according to elapsed time.
        var now = FoptNow(_options)();
        var elapsed = now - _last;
        _last = now;
        _tokens = Math.Min(burst, _tokens + (elapsed / 1000.0) * rate);

        if (_tokens >= 1)
        {
            _tokens -= 1;
            return;
        }

        // Not enough tokens: wait for one to accrue, then consume it.
        var needed = 1 - _tokens;
        var waitMs = (int)Math.Ceiling((needed / rate) * 1000);
        Track(waitMs);
        if (waitMs > 0)
        {
            FoptSleep(_options)(waitMs);
        }
        _last = FoptNow(_options)();
        _tokens = 0;
    }

    private void Track(int waitMs)
    {
        Throttled++;
        WaitMs += waitMs;
    }
}

File: ts/project/.sdk/tm/go/feature/ratelimit_feature.go
package feature

import (
	"math"

	"GOMODULE/core"
)

// Client-side rate limiting via a token bucket. Each request consumes a
// token; when the bucket is empty the request waits until the bucket
// refills at `rate` tokens per second (with capacity `burst`, default:
// rate). This keeps the client under a server's published quota rather
// than discovering it via 429s. The clock (`now`) and the wait (`sleep`)
// are injectable so the accounting can be tested deterministically.
type RatelimitFeature struct {
	BaseFeature
	client  *core.ProjectNameSDK
	options map[string]any
	tokens  float64
	last    int64

	// Activity tracking (mirrors the ts client._ratelimit record).
	Throttled int
	WaitMs    int
}

func NewRatelimitFeature() *RatelimitFeature {
	return &RatelimitFeature{
		BaseFeature: BaseFeature{
			Version: "0.0.1",
			Name:    "ratelimit",
			Active:  true,
		},
	}
}

func (f *RatelimitFeature) Init(ctx *core.Context, options map[string]any) {
	f.client = ctx.Client
	f.options = options
	f.Active = foptBool(options, "active", false)

	if !f.Active {
		return
	}

	rate := foptNum(f.options, "rate", 5)
	burst := foptNum(f.options, "burst", rate)
	f.tokens = burst
	f.last = foptNow(f.options)()

	inner := ctx.Utility.Fetcher

	ctx.Utility.Fetcher = func(ctx2 *core.Context, url string, fetchdef map[string]any) (any, error) {
		f.acquire()
		return inner(ctx2, url, fetchdef)
	}
}

func (f *RatelimitFeature) acquire() {
	rate := foptNum(f.options, "rate", 5)
	burst := foptNum(f.options, "burst", rate)

	// Refill according to elapsed time.
	now := foptNow(f.options)()
	elapsed := now - f.last
	f.last = now
	f.tokens = math.Min(burst, f.tokens+(float64(elapsed)/1000)*rate)

	if f.tokens >= 1 {
		f.tokens -= 1
		return
	}

	// Not enough tokens: wait for one to accrue, then consume it.
	needed := 1 - f.tokens
	waitMs := int(math.Ceil((needed / rate) * 1000))
	f.track(waitMs)
	if waitMs > 0 {
		foptSleep(f.options)(waitMs)
	}
	f.last = foptNow(f.options)()
	f.tokens = 0
}

func (f *RatelimitFeature) track(waitMs int) {
	f.Throttled++
	f.WaitMs += waitMs
}
